frontend/src/general/SnackbarMessage.jsx

```diff
@@ -0,0 +1,166 @@
+  }
+
+  handleClose(event, reason) {
+    if (reason === "clickaway") {
+      return;
+    }
+    this.setOpen(false);
+  }
+
+  setOpen(open) {
+    this.setState({
+      open
+    });
+  }
+
+  setStatus(status) {
+    this.setState({
+      status
+    });
+  }
+
+  setIsSuccess(isSuccess) {
+    this.setState({
+      isSuccess
+    });
+  }
+
+  setMessage(message) {
+    this.setState({
+      message
+    });
+  }
+
+  setPathname(pathname) {
+    this.setState({
+      pathname
+    });
+    return pathname;
+  }
+
+  addSnackbarProperties(stat, msg) {
+    this.setStatus(stat);
+    if (stat !== "error") {
+      this.setIsSuccess(true);
+    } else {
+      this.setIsSuccess(false);
+    }
+    this.setMessage(msg);
+  }
+
+  openSnackbar(stat, msg, path) {
+    const currentPathname = this.setPathname(path);
+    if (currentPathname !== "") {
+      this.refresh();
+      setTimeout(() => {
+        this.addSnackbarProperties(stat, msg);
+        this.setOpen(true);
+      }, 300);
+    } else {
+      this.addSnackbarProperties(stat, msg);
+      this.setOpen(true);
+    }
+  }
+
+  refresh() {
+    const { history, location } = this.props;
+    const { pathname } = this.state;
+    if (location.pathname.endsWith(pathname)) {
+      history.go(0);
+    } else {
+      history.push(`/${pathname}`);
+    }
+  }
+
+  render() {
+    const { open, isSuccess, status, message } = this.state;
+    return (
+      <>
+        <Snackbar
+          contentprops={{
+            "aria-describedby": "message-id"
+          }}
+          anchorOrigin={{
+            vertical: "top",
+            horizontal: "right"
+          }}
+          open={open}
+          autoHideDuration={3000}
+          onClose={this.handleClose}
+        >
+          <SnackbarContent
+            style={{
+              backgroundColor: isSuccess ? "#4caf50" : "#f44336"
+            }}
+            // prettier-ignore
+            message={(
+              <Box id='message-id' className="snackMessage">
+                {isSuccess ? 
+                  <CheckCircle className="iconPadding" /> : 
+                  <Error className="iconPadding" />}
+                {message || `Form submission status: ${status}`}
+              </Box>
+            )}
+            action={[
+              <IconButton
+                key="close"
+                aria-label="Close"
+                color="inherit"
+                onClick={this.handleClose}
+              >
+                <Close />
+              </IconButton>
+            ]}
+          />
+        </Snackbar>
+      </>
+    );
+  }
+}
+
+SnackbarMessage.propTypes = {
+  history: PropTypes.shape({
+    push: PropTypes.func.isRequired,
+    go: PropTypes.func.isRequired
+  }).isRequired,
+  location: PropTypes.shape({
+    pathname: PropTypes.string.isRequired
+  }).isRequired
+};
+
+export function snackbar(status, message, pathname = "") {
+  openSnackbarFn(status, message, pathname);
+}
```

Comment:
So from what I can tell, this is how snackbar works

- One SnackbarMessage component exists in AppRouter (so it exists in all pages)
- Any component can import this `snackbar` global function
- When calling `snackbar` it opens the snackbar from anywhere, with the given status and message

Usually I would say exposing a global function like this is kind of an antipattern, but I guess the only other way to do it is to pass the snackbar function along in all the props of every component, which would be less clean.

My question is: is this the "standard" for how snackbar opening is done? Do you have an article or something that you followed for this design? I think it would be a useful reference for me to trust that this is a tried-and-true architecture decision.

Reply:
https://www.freecodecamp.org/news/how-to-show-informational-messages-using-material-ui-in-a-react-web-app-5b108178608/

https://github.com/async-labs/builderbook/blob/master/book/4-end/components/Notifier.jsx

Reply:
Thank you!!